dbms/src/WindowFunctions/tests/gtest_lead_lag.cpp

```diff
@@ -0,0 +1,435 @@
+         toNullableVec<Int64>(/*order*/ {1, 2, 3, 4, 5, 6, 7, 8}),
+         toNullableVec<String>(/*value*/ {"1", "2", "3", "4", "5", "6", "7", "8"})});
+
+    // arg2 < partition_size
+    executeFunctionAndAssert(
+        toNullableVec<String>({"2", "3", "4", "0", "6", "7", "8", "0"}),
+        Lead3(value_col, lit(Field(static_cast<UInt64>(1))), lit(Field(String("0")))),
+        {toNullableVec<Int64>(/*partition*/ {1, 1, 1, 1, 2, 2, 2, 2}),
+         toNullableVec<Int64>(/*order*/ {1, 2, 3, 4, 5, 6, 7, 8}),
+         toNullableVec<String>(/*value*/ {"1", "2", "3", "4", "5", "6", "7", "8"})});
+    executeFunctionAndAssert(
+        toNullableVec<String>({"0", "1", "2", "3", "0", "5", "6", "7"}),
+        Lag3(value_col, lit(Field(static_cast<UInt64>(1))), lit(Field(String("0")))),
+        {toNullableVec<Int64>(/*partition*/ {1, 1, 1, 1, 2, 2, 2, 2}),
+         toNullableVec<Int64>(/*order*/ {1, 2, 3, 4, 5, 6, 7, 8}),
+         toNullableVec<String>(/*value*/ {"1", "2", "3", "4", "5", "6", "7", "8"})});
+
+    // arg2 >= partition_size
+    executeFunctionAndAssert(
+        toNullableVec<String>({"0", "0", "0", "0", "0", "0", "0", "0"}),
+        Lead3(value_col, lit(Field(static_cast<UInt64>(4))), lit(Field(String("0")))),
+        {toNullableVec<Int64>(/*partition*/ {1, 1, 1, 1, 2, 2, 2, 2}),
+         toNullableVec<Int64>(/*order*/ {1, 2, 3, 4, 5, 6, 7, 8}),
+         toNullableVec<String>(/*value*/ {"1", "2", "3", "4", "5", "6", "7", "8"})});
+    executeFunctionAndAssert(
+        toNullableVec<String>({"0", "0", "0", "0", "0", "0", "0", "0"}),
+        Lead3(value_col, lit(Field(static_cast<UInt64>(5))), lit(Field(String("0")))),
+        {toNullableVec<Int64>(/*partition*/ {1, 1, 1, 1, 2, 2, 2, 2}),
+         toNullableVec<Int64>(/*order*/ {1, 2, 3, 4, 5, 6, 7, 8}),
+         toNullableVec<String>(/*value*/ {"1", "2", "3", "4", "5", "6", "7", "8"})});
+    executeFunctionAndAssert(
+        toNullableVec<String>({"0", "0", "0", "0", "0", "0", "0", "0"}),
+        Lag3(value_col, lit(Field(static_cast<UInt64>(4))), lit(Field(String("0")))),
+        {toNullableVec<Int64>(/*partition*/ {1, 1, 1, 1, 2, 2, 2, 2}),
+         toNullableVec<Int64>(/*order*/ {1, 2, 3, 4, 5, 6, 7, 8}),
+         toNullableVec<String>(/*value*/ {"1", "2", "3", "4", "5", "6", "7", "8"})});
+    executeFunctionAndAssert(
+        toNullableVec<String>({"0", "0", "0", "0", "0", "0", "0", "0"}),
+        Lag3(value_col, lit(Field(static_cast<UInt64>(5))), lit(Field(String("0")))),
+        {toNullableVec<Int64>(/*partition*/ {1, 1, 1, 1, 2, 2, 2, 2}),
+         toNullableVec<Int64>(/*order*/ {1, 2, 3, 4, 5, 6, 7, 8}),
+         toNullableVec<String>(/*value*/ {"1", "2", "3", "4", "5", "6", "7", "8"})});
+
+    // test arg column type
+    executeFunctionAndAssert(
+        toNullableVec<String>({"2", "3", "4", "4", "6", "7", "8", "8"}),
+        Lead3(value_col, lit(Field(static_cast<UInt64>(1))), value_col),
```

Comment:
Any test to test the case that 1st and 3rd argument has different type?

Reply:
It is difficult to add such cases in UT... Because it requires type derivation in TestUtils.
https://github.com/pingcap/tiflash/pull/5671/files#diff-d3bab376515e4a6b7c7170937d4ad7517191427f9c30103f12299707794c1f66R1897-R1911
https://github.com/pingcap/tiflash/pull/5671/files#diff-d3bab376515e4a6b7c7170937d4ad7517191427f9c30103f12299707794c1f66R1443-R1465
Now only the case where 1st and 3rd argument have the same type is supported..
But I added these cases to `tests/fullstack-test/expr/lead_lag.test`.
[lead(value_decimal32, 1, 'default')](https://github.com/pingcap/tiflash/pull/5671/files#diff-659a917bbb1b8007e0788ef8fe19b95530a44700e97e923cb6ddac6cccd047c4R144)